commitrmnocb/outcome.go

```diff
@@ -0,0 +1,325 @@
+		if f, exists := fChains[chain]; exists {
+			root, count := mostFrequentElem(roots)
+
+			if count <= f {
+				// TODO: metrics
+				p.lggr.Warnf("failed to reach consensus on a merkle root for chain %d "+
+					"because no single merkle root was observed more than the expected %d times, found merkle root %d "+
+					"observed by only %d oracles, all observed merkle roots: %v",
+					chain, f, root, count, roots)
+			}
+
+			consensus[chain] = root
+		} else {
+			// TODO: metrics
+			p.lggr.Warnf("merkleRootConsensus: fChain not found for chain %d", chain)
+		}
+	}
+
+	return consensus
+}
+
+// Given a mapping from chains to a list of max seq nums on their corresponding OnRamp, return a mapping from chains
+// to a single max seq num. The consensus max seq num for a given chain is the f'th lowest max seq num if the number
+// of max seq num observations is greater or equal than 2f+1, where f is the FChain of the corresponding source chain.
+func (p *Plugin) onRampMaxSeqNumsConsensus(
+	onRampMaxSeqNumsByChain map[cciptypes.ChainSelector][]cciptypes.SeqNum,
+	fChains map[cciptypes.ChainSelector]int,
+) map[cciptypes.ChainSelector]cciptypes.SeqNum {
+	consensus := make(map[cciptypes.ChainSelector]cciptypes.SeqNum)
+
+	for chain, onRampMaxSeqNums := range onRampMaxSeqNumsByChain {
+		if f, exists := fChains[chain]; exists {
+			if len(onRampMaxSeqNums) < 2*f+1 {
+				// TODO: metrics
+				p.lggr.Warnf("could not reach consensus on onRampMaxSeqNums for chain %d "+
+					"because we did not receive more than 2f+1 observed sequence numbers, 2f+1: %d, "+
+					"len(onRampMaxSeqNums): %d, onRampMaxSeqNums: %v",
+					chain, 2*f+1, len(onRampMaxSeqNums), onRampMaxSeqNums)
+			} else {
+				sort.Slice(onRampMaxSeqNums, func(i, j int) bool { return onRampMaxSeqNums[i] < onRampMaxSeqNums[j] })
+				consensus[chain] = onRampMaxSeqNums[f]
+			}
+		} else {
+			// TODO: metrics
+			p.lggr.Warnf("could not reach consensus on onRampMaxSeqNums for chain %d "+
+				"because there was no consensus f value for this chain", chain)
+		}
+	}
+
+	return consensus
+}
+
+// Given a mapping from chains to a list of max seq nums on the OffRamp, return a mapping from chains
+// to a single max seq num. The consensus max seq num for a given chain is the max seq num with the most observations
+// that was observed at least f times, where f is the FChain of the dest chain.
+func (p *Plugin) offRampMaxSeqNumsConsensus(
+	offRampMaxSeqNumsByChain map[cciptypes.ChainSelector][]cciptypes.SeqNum,
+	fDestChain int,
+) map[cciptypes.ChainSelector]cciptypes.SeqNum {
+	consensus := make(map[cciptypes.ChainSelector]cciptypes.SeqNum)
+
+	for chain, offRampMaxSeqNums := range offRampMaxSeqNumsByChain {
+		seqNum, count := mostFrequentElem(offRampMaxSeqNums)
+		if count <= fDestChain {
+			// TODO: metrics
+			p.lggr.Warnf("could not reach consensus on offRampMaxSeqNums for chain %d "+
+				"because we did not receive a sequence number that was observed by at least f (%d) oracles, "+
```

Comment:
```suggestion
				"because we did not receive a sequence number that was observed by at least fChain (%d) oracles, "+
```

To make it clear this is the fChain and not big F value